ecs-cli/modules/cli/local/project/project.go

```diff
@@ -0,0 +1,214 @@
+// Copyright 2015-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Package localproject defines LocalProject interface and implements them on localProject
+
+package localproject
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/cli/local/converter"
+	ecsclient "github.com/aws/amazon-ecs-cli/ecs-cli/modules/clients/aws/ecs"
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/clients/aws/secretsmanager"
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/commands/flags"
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/config"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/urfave/cli"
+)
+
+const (
+	LocalOutDefaultFileName = "./docker-compose.local.yml"
+	LocalOutFileMode        = os.FileMode(0644) // Owner=read/write, Other=readonly
+	LocalInFileName         = "./task-definition.json"
+)
+
+type LocalProject interface {
+	ReadTaskDefinition() error
+	Convert() error
+	Write() error
+	LocalOutFileName() string
+}
+
+type localProject struct {
+	context          *cli.Context
+	taskDefinition   *ecs.TaskDefinition
+	localBytes       []byte
+	localOutFileName string
+}
+
+func New(context *cli.Context) LocalProject {
+	p := &localProject{context: context}
+	return p
+}
+
+func (p *localProject) TaskDefinition() *ecs.TaskDefinition {
+	return p.taskDefinition
+}
+
+func (p *localProject) LocalOutFileName() string {
+	return p.localOutFileName
+}
+
+func (p *localProject) ReadTaskDefinition() error {
+	arn := p.context.String(flags.TaskDefinitionArnFlag)
+	filename := p.context.String(flags.TaskDefinitionFileFlag)
+
+	if arn != "" && filename != "" {
+		return fmt.Errorf("Cannot specify both --%s and --%s flags.", flags.TaskDefinitionArnFlag, flags.TaskDefinitionFileFlag)
```

Comment:
Hm, I'm not sure. As clunky and verbose as it is, I *thought* it was idiomatic to bubble up all the errors to the main caller, who would have the responsibility of deciding whether to  use `logrus.Fatal`. This also seems to be how we do it in the compose workflows, and while it's not necessarily done right there this would at least be consistent. I think I am hesitant to make these methods crash the app since that might make it very restrictive for re-use in other callers?

Reply:
I agree that it's a more aggressive way of handling the error instead of just returning it, however I just don't see how else these errors could be handled. I don't think it's idiomatic go to bubble up errors (https://dave.cheney.net/2019/01/27/eliminate-error-handling-by-eliminating-errors), but I can't find good guidelines either :/  

I like following the idea of reducing the number of exceptions that must be handled. Since for any of these errors the only possible action is to crash, I'd rather do that where the exception happens instead of the caller.

I'm fine with keeping it as is though since it's still readable, it just seems unnecessary boilerplate in my opinion.